### ISSUE TITLE
Question: how to drive futures forward

### ISSUE THREAD

**Issue by orthecreedence** [alexcrichton/futures-rs#332]
I have an app with a main thread that waits on incoming messages from other pieces of the app (separate threads). Essentially, it acts as an event loop, blocking on waiting for incoming messages on its channel.

With older versions of futures-rs, I could use the `.forget()` method to spawn a future on this main thread then have the main thread continue to execute it as it progressed. It seems that now the only simple way to drive a future forward is by spawning a new thread and running `future.wait()` inside of it. I'd really like to avoid doing this but have not found a way to drive a future forward on the main thread.

I tried looking at the Tokio implementation but couldn't make much sense of it.

I've tried calling `future.poll()` but it panics and says there's no Task. I cannot find a way to create a Task object.

Is there a way to mimic how `future.forget()` worked in the past with the latest version?

**Comment by carllerche** [comment]
Now, you would need to explicitly spawn the future on a thread pool. The `forget` fn had some unintended implications, so it was removed.

The easiest way right now would be to use futures-cpupool and spawn the future on that.

**Comment by orthecreedence** [comment]
So there's no public way to drive a future forward step-by-step on a specific thread?

**Comment by carllerche** [comment]
@orthecreedence well, "on a specific thread" would just be "Future::wait()" on the thread you want. `forget()` did something worse... it called poll() immediately on whichever thread happened to "notify" the task. This meant that it could have some pretty bad unintended side effects. 

Maybe you could describe a bit more what behavior you are looking for exactly and I could try to tell you how to do it.

**Comment by dwrensha** [comment]
@orthecreedence you can drive a future step-by-step by first calling [`::futures::executor::spawn()`](http://alexcrichton.com/futures-rs/futures/executor/fn.spawn.html) and then calling [`poll_future()`](http://alexcrichton.com/futures-rs/futures/executor/struct.Spawn.html#method.poll_future) on the returned handle, providing an `Unpark` implementation that ignores all notifications. Or if you don't care about step-by-step, you could call [`wait_future()`](http://alexcrichton.com/futures-rs/futures/executor/struct.Spawn.html#method.wait_future), which is what `Future::wait()` uses under the hood.

**Comment by orthecreedence** [comment]
I did not know that calling `poll_future()` drove things forward. I think I tried this, but perhaps didn't wire it up correctly. I'll play around with this for a few minutes. So calling `executor::spawn()` takes a future and creates a new task for it (bundled together in a `Spawn`)? Then calling `spawn.poll_future()` drives this task forward one step?

Is there an example implementation for an `Unpark` that does what you're saying? Like the quickest way to get it running? I'll poke around the source and see if I can find something.

Thank you both for your help. If it wasn't obvious, even after reading the docs around six times, I'm not very clear on what a Task/Executor is and how they fit together, so I feel like I'm kind of bumbling through this. things are getting more clear as I'm playing around though.

**Comment by orthecreedence** [comment]
Ok, I tried using `poll_future()` but it doesn't seem to be driving my futures forward. Maybe if I give some background I won't seem so annoying about this issue.

My use case is this. I have kind of a makeshift event loop:

```rust
// app state passed to every callback run by main
let app_state = make_app_state();
// make an mpsc channel our main thread grabs functions from
let (tx, rx) = make_channel();
// main thread loop
loop {
    let fn = rx.recv();
    fn.call(app_state.clone())
}
```

So the main thread inits its state then runs in a loop, receiving messages (really, callbacks) from other threads. What I really want is for futures created on the main thread to be driven forward by the main thread so that any non-cpu-intensive work is happening on the main thread (and of course, offloading intense stuff to a CpuPool).

I'd want something like this:

```rust
let future = my_async_op()
    .then(|x| {
        my_other_async_op()
    });
// drives the future forward
drive_future(future, main_tx);
```

The way `forget()` worked, it would start the future on the main thread, then when something async occurred, I could create a oneshot future, wait for the async op to complete, then complete the oneshot via something like `tx_main.send(|_| oneshot_tx.complete(result); )` if that makes sense. So I'm trying to recreate this flow because I'm somewhat partial to all non-cpu work running on the main thread (due to it being the "event loop" thread).

So really, I'm trying to mimic `.forget()` in a sense, even if it was incorrect, because it did what I wanted it to. Thanks again for the help!

**Comment by orthecreedence** [comment]
I think I'm figuring things out, please forgive the noise. I'm going to close this issue for now.
